### PR TITLE
feat(backend): 支持普通与服务草稿 Bot 的 UPFS 创建和普通重启

### PR DESCRIPTION
**Pull request by maicaisongxiaocong**
## Problem

新建 Bot 需要按灰度使用 UPFS；普通 Bot 与服务草稿普通重启需要保留已有存储选择，不能在重启时重新灰度。PR 原有路由文件达到 1005 行导致 Backend CI 失败，评审另指出环境读取、运行时名称分支、重复能力检查和插件契约测试的问题。

## Solution

- 支持普通 Bot/服务 Draft 创建及普通重启；创建共用独立 `initialize_bot_storage → BotStoragePolicyService.initialize`，普通重启只读已有策略，走已有 BaaS Bot `/update`，不重新分配实例。
- 新 Bot 的 NAS/UPFS 首次选择都保存到唯一 `storage_policy`；旧 Bot 无策略保持 NAS，不补写、不迁移。本次仅保存 `storage_type/source`，不包含已暂缓的创建状态、失败原因或 CAS 结果回写。
- Backend 不访问 BaaS 表、不获取 Volume ID，仅调用安全模板能力接口；BaaS 查询当前环境 Volume、生成挂载参数并在正式创建前再校验。SDK 升级和企业适配由 OCB 主仓配套交付。
- 按评论修复：部署组件通过 `supports_bot_storage_policy` 声明能力；BaaS 模板服务接收 bootstrap 注入的环境，Facade 复用环境快照；三处 UPFS 模板检查收口；增加四种社区插件的支持/拒绝契约测试。
- 路由删去重复参数说明降至 983 行，保留原有 1000 行 CI 门禁，不增加豁免。同步仓库文档和语雀当前实现范围，删除旧的已实现状态回写说明。

## Validation

- 创建/真实重启入口专项与文件大小门禁：90 passed。
- Backend 配置、仓储、设备、服务 Bot、Bot 管理、DI 和相关架构回归：3299 passed；3 个此前复现的本地环境问题（缺 pymysql、AWS 代理、企业模块可见性）。
- 公共 BaaS unit/architecture：9054 passed、3 skipped、8 xpassed；本地排除 1 项既有全仓格式检查，未修改 CI 或豁免。本轮改动文件 Ruff/格式检查通过。
- 企业 BaaS unit/contract：656 passed、15 skipped。
- Ruff、diff、推送前密钥/SAST 检查通过。GitHub 对 `d2a2ca8f0` 的 CI 正在重跑，最终结果以该提交的 checks 为准。
- 未执行真实 ARCA 挂载、生产 DDL、生产配置或数据迁移。

## Compatibility and risk

仅覆盖已有实例普通重启；不承诺 Bot/Device 丢失后重建兜底的数据保留。完整 Verify/Online 发布、已发布服务版本重启/升级、恢复草稿、个人转服务编排、扩容和 Caller 尚未交付。

新建 NAS 保存策略是本分支行为，不代表线上已部署；存量无策略 NAS 不补记录。关闭灰度不改变已保存策略，不能删除策略作为数据回滚。

企业 `arca-sandbox==1.3.0`、SDK 转换和部署 DDL 位于 OCB 主仓，本公共提交不包含企业代码。

## Spec

`docs/specs/2026-09-15-upfs-phase1/spec.md`；语雀设计已通过 CLI 同步当前范围，§4.1.1 明确为后续阶段。

### CODE REVIEW

**Review by totalfrank — APPROVED** [review]
## Architecture-Gate Review

**Verdict: APPROVE**

### Gate 0
- Target branch: `dev` (the integration branch) — not a `REL*` release branch, so the release-branch gate and Isolation Audit do not apply.
- Classification: genuine feature work landing on its normal integration branch. Pass.

### Review summary

Read `docs/arch/arch.rules.md` and `AGENTS.md`, then the full diff (47 files, ~3.1k additions) across `src/baas/src/secbaas/...` and `src/backend/src/agentclaw/...`. Findings by architectural role:

- **New Service API contracts** (`BotCommonConfigServiceProtocol`, `BotStoragePolicyProtocol` in `agentclaw/community/core/common_config/bot_config_protocol.py`; `DeviceTemplateManageService.get_storage_capability` in `secbaas/community/api/template_manage/_protocols.py`) are defined as explicit Protocols in the engine-agnostic layer, separate from delivery (Rule 3/5). Implementations (`BotCommonConfigService`, `BotStoragePolicyService`, `BotCommonConfigRepository`) live in `core/` and are wired only from composition roots (`di/modules/common_config_module.py`, `di/modules/service_bot_module.py`, `secbaas/bootstrap/_container.py` + `_core_services.py`) — Rule 14 respected, no implementation selection leaks into business code.
- **Plugin API extension is additive and conformant.** `ArcaSandboxPlugin.create_sync_sandbox` gains one new optional `volume_mounts: list[VolumeMountSpec] | None = None` parameter (`secbaas/community/spi/sandbox/arca/_protocols.py`). Every existing implementation is updated in the same PR: `StubArcaSandboxPlugin` supports it, `AliyunAckSandboxPlugin`/`LocalDockerArcaSandboxPlugin`/`LocalProcessArcaSandboxPlugin` explicitly `raise NotImplementedError` when mounts are supplied. This is a real system-boundary plugin (external Arca SDK), so Plugin API is the right mechanism (not the engine-contract pattern), and the "unsupported implementations must reject them" semantics are documented directly on the protocol (Rule 25 conformance).
- **Engine variability goes through the existing contract, not new branching.** `baas_device_service.py` resolves the rollout bucket via the pre-existing `resolve_baas_engine_bucket()` composition-root registry (`core/bot_management/engines/registry.py`, untouched by this PR) rather than adding `if engine_type == "..."` checks. `should_rollout_upfs()` (`core/common_config/upfs_rollout.py`) is a pure function that takes `engine`/`user_id` and matches against externally supplied config (`engine_bucket` rules from `ac_common_config`) — no engine names are hardcoded in domain code. Applied the "add engine `foo`" test: a new engine only needs a registry entry in the existing registry plus its own rollout rule row; zero edits to `bot_storage_policy.py` or `upfs_rollout.py` would be needed.
- **Contract/DTO changes are additive and backward-compatible.** `ArcaTemplateConfig.upfs_volume_id_pre/prod`, `ArcaCreateConfig.volume_mounts`, `ArcaDeviceConfig.volume_mounts`, `BotDeployContext.storage_type`, `StorageType.UPFS` — all new optional fields/enum members with safe defaults (NAS-compatible). `VolumeMountSpec` validates its own inputs (path traversal, absolute mount path, non-empty ids) at the contract boundary.
- **Config-driven, not hardcoded policy.** The UPFS subpath quota reads `SystemConfigKey.UPFS_SUBPATH_SIZE_BYTES` via `SystemConfigManageService`; the `1073741824` literal is only a documented fallback when config is missing/invalid, not the operative value.
- **DB migration is additive.** `2026_09_15_bot_common_config.sql` is a single `CREATE TABLE IF NOT EXISTS` with a new unique key — no destructive or rewriting migration.
- **Propagation analysis is present.** `core/common_config/README.md` documents the context boundary (provides/consumes/internal_dependencies), a "Change impact" section, the phase-one contract, the persisted JSON shape, and an explicit list of what is *not* yet supported (service Verify/Online, build/restore, Caller, scale). This satisfies Rule 16's propagation requirement for the new Service API and repository contract.
- **Tests**: `tests/community/core/common_config/test_bot_storage_policy.py` (1050 lines) exercises the repository/service contracts against a real SQLite-backed implementation (concurrency/CAS included), not just mocks; `src/baas/tests/unit/core/service/device_manage/test_upfs_storage.py` covers the ARCA-only gate and volume mount construction.

### Findings

None. No engine-string branching, hardcoded capability sets, engine-named modules under `core/`, raw `os.environ`/`os.getenv` additions, transport/framework leakage into core, or plugin-vs-engine misuse were introduced or extended by this diff.

### Out-of-scope notes

- `secbaas/community/core/service/template_manage/_device_template_service.py` newly imports `get_current_env()` for the `storage-capability` preflight check. This mirrors ~8 pre-existing call sites of the same utility in the sibling `_device_service.py` in the same module, so it's consistent with this service's established (if debatable) convention rather than a new pattern — flagging for awareness, not blocking.
- Business/product semantics (rollout targeting rules, retry/CAS bookkeeping correctness, "personal vs service Draft" scoping) were assumed correct per task instructions and not re-reviewed here.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by totalfrank — CHANGES_REQUESTED** [review]
## 1. Verdict: REQUEST_CHANGES

## 2. Gate 0

- **Target branch:** `dev` (the integration branch) — this is the normal route, not a `REL*` release branch, so the release-branch gate and Isolation Audit do not apply.
- **Classification:** Feature work extending existing pipelines (BaaS device provisioning, service-bot deploy composition, bot common-config) — appropriately targeted at `dev`.
- **Pass/fail:** Pass. No Gate 0 issue.

## 3. Findings

**1. `core/` reaches into environment detection directly — new call site, Rule 14 (Invariant) / Structural Red Flag "environment access scattered outside config or bootstrap"**

`src/baas/src/secbaas/community/core/service/template_manage/_device_template_service.py:30,170`

```python
from secbaas.community.core.utils.env_utils import get_current_env
...
    def get_storage_capability(...) -> TemplateStorageCapability:
        current_env = get_current_env()
```

`_device_template_service.py` had zero calls to `get_current_env()` before this PR — this diff is the one introducing environment/mode detection into this specific core file. `env_utils.get_current_env()` itself does `os.getenv`/config-file reads (`core/utils/env_utils.py`), i.e. exactly the "environment access" Rule 14 and the constitution's red-flag list reserve for composition roots/bootstrap, not core services. Yes, the sibling file `_device_service.py` already calls `get_current_env()` at eight sites — but per the review ground rules, a pre-existing violation nearby is not a license to add a new call site; it makes this an extension of the same violation, which is itself a finding.

Concretely, this PR already knows the compliant shape and used it correctly on the other side of the same feature: `di/modules/service_bot_module.py` injects `storage_env=env_utils.get_current_env()` into `BaasService.__init__` from the composition root, and the service just reads `self._storage_env`. `DefaultDeviceTemplateService` should receive the current deployment env the same way — as a constructor parameter wired in `bootstrap/_core_services.py`/`_container.py` — instead of calling `get_current_env()` from inside `get_storage_capability`. This keeps the service transport/environment-agnostic and testable without monkeypatching `env_utils`.

## 4. Out-of-scope notes

- `src/backend/tests/community/core/common_config/test_bot_storage_policy.py` is exactly 1050 lines, over AGENTS.md's stated 1000-line CI threshold. Flagging only as a heads-up (test file, not gated by this review's scope).
- `BotStoragePolicyProtocol` (`bot_config_protocol.py`) declares `_resolve_storage_policy` (leading underscore) as part of the public contract, and it's called cross-module from `core/service_bot/services/baas_service.py`. Naming/contract-hygiene nit, not gated here.
- Business logic (UPFS rollout rule semantics, storage-policy JSON shape, retry/CAS behavior) looked correct and is well-covered by tests and the new `common_config/README.md` context-boundary doc; not reviewed for correctness per this review's scope.
- Engine variability (aicoding/claude_code/openclaw/teclaw) introduced by this PR (`resolve_baas_engine_bucket` call site, `should_rollout_upfs`'s `engine_bucket` matching) correctly stays config/registry-driven with no new `if engine == ...` branching in `core/` — called out as a positive, not a finding.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/baas_service.py [hunk 1]
```diff
@@ -629,6 +639,71 @@ def _resolve_service_bot_resource_spec(
...
 
+    def begin_bot_storage_creation(
+        self, *, bot_id: str, entity_id: str, env: str
+    ) -> str | None:
+        """Track managed personal/service Draft allocations with an existing policy."""
+        from agentclaw.community.kernel.deploy_runtime import DeployRuntime
+
+        if (
+            self._bot_storage_service is None
+            or self._deploy_composer.name != DeployRuntime.MANAGED
```
Comment (totalfrank) [inline src/backend/src/agentclaw/community/core/service_bot/services/baas_service.py]: **[Architecture] Under-abstraction: branching on `DeployConfigComposer` identity inside a core service — arch.rules.md Rule 14 (Invariant), also violates this module's own documented contract**

```python
if (
    self._bot_storage_service is None
    or self._deploy_composer.name != DeployRuntime.MANAGED
):
    return None
```

Same pattern repeats at `initialize_bot_storage` (line 683): `self._deploy_composer is None or self._deploy_composer.name != DeployRuntime.MANAGED`.

`BaasService` is `core/` and `DeployConfigComposer` is exactly the swappable-strategy contract this constitution requires (Rule 10) — but `deploy_config_composer.py`'s own docstring for `.name` says: *"Which composer runs is a deployment's choice, made once in the composition root from validated config (Rule 14) — **never branched on per request**."* This PR does exactly that: it inspects the injected composer's identity (`DeployRuntime.MANAGED`) inline in `BaasService` to decide whether the new bot-storage-policy feature applies, rather than asking the composer itself.

This is the same "under-abstraction" failure mode the review guide calls out for engine-string branching, generalized to a deploy-runtime discriminator: apply the "add a new runtime" test — today, wiring a new `DeployConfigComposer` implementation that should also support storage-policy tracking requires editing this `if` in `BaasService`, not just adding/registering the new composer. That is the acid-test failure condition (an engine-agnostic/runtime-agnostic core file needing edits for a new implementation).

**Fix direction:** expose the capability on the contract instead of inspecting its identity, e.g. `DeployConfigComposer.supports_bot_storage_policy: bool` (default `False` on the ABC or explicit per composer — `True` only on `ManagedDeployConfigComposer`), and have `BaasService` call `self._deploy_composer.supports_bot_storage_policy` here and at line 683. `ManagedDeployConfigComposer` already declares `name -> DeployRuntime.MANAGED` as its one authorized identity field for logging/config-drift checks (per its own docstring) — it should not also become the thing core branches on.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/baas/src/secbaas/community/core/service/template_manage/_device_template_service.py [hunk 2]
```diff
@@ -159,6 +161,45 @@ def __init__(
         self._tenant_service = tenant_service
         self._secret_plugin = secret_plugin
 
+    def get_storage_capability(
+        self,
+        tenant: str,
+        template_uuid: str,
+        env: str | None = None,
+    ) -> TemplateStorageCapability:
+        current_env = get_current_env()
```
Comment (totalfrank) [inline src/baas/src/secbaas/community/core/service/template_manage/_device_template_service.py]: **[Architecture] Environment detection inside `core/` — arch.rules.md Structural Review Red Flags ("environment access is scattered outside config or bootstrap"), catalog item 6**

```python
current_env = get_current_env()
```

This is a new call site (new `from secbaas.community.core.utils.env_utils import get_current_env` import at line 30) in a file that previously never read the runtime environment directly — it's used here to validate the caller-supplied `env` against "where BaaS is actually running" for the new `get_storage_capability` preflight endpoint.

For transparency: this pattern is already pervasive elsewhere in `secbaas/community/core/` (`_device_service.py` alone has 8 pre-existing call sites, and it appears in ~38 files repo-wide), so this isn't inventing a new mechanism — but per the review's own rule, extending an existing violation into a file that didn't have it is still a violation, not grandfathered by precedent. `core/` should receive its environment as a value (constructor/method parameter resolved once at the composition root/bootstrap), not go looking for it itself.

**Fix direction:** either thread the current environment in from the router/composition root as a parameter (consistent with how `env: str | None` is already accepted from the caller here), or — if this pervasive pattern is intentionally tolerated for this component — file a written waiver per the Governance Addendum (rule, reason, risk, compensating controls, owner, expiry) rather than continuing to add call sites silently.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by totalfrank — CHANGES_REQUESTED** [review]
## Architecture-Gate Review — REQUEST_CHANGES

**Scope note:** this review is structural only (per `docs/arch/arch.rules.md` and `AGENTS.md`) — the NAS→UPFS rollout product logic, its safety guards, and the extensive test/doc coverage are not evaluated here and look thorough at a glance.

### Gate 0 — target branch
Target is `dev` (the integration branch), not a `REL*` release branch, so the release-branch isolation gate and its restrictions do not apply. Gate 0: **pass**.

### Findings (most severe first)

1. **`src/backend/.../core/service_bot/services/baas_service.py:650` and `:683`** — `BaasService.begin_bot_storage_creation` / `initialize_bot_storage` branch on `self._deploy_composer.name != DeployRuntime.MANAGED` to decide whether the new storage-policy feature runs. `DeployConfigComposer` is exactly the swappable-strategy contract Rule 10 calls for, and its own docstring states the composer's identity is chosen once in the composition root and "never branched on per request." This PR does that branching inline in a `core/` service — the generalized form of the engine-string-branching anti-pattern this gate hunts for (Rule 14, Invariant). Fix: add a capability to the `DeployConfigComposer` contract (e.g. `supports_bot_storage_policy`) that `ManagedDeployConfigComposer` implements as `True`, and have `BaasService` ask the composer instead of inspecting its `.name`. See inline comment for detail.

2. **`src/baas/.../core/service/template_manage/_device_template_service.py:170`** — new `get_current_env()` call site added inside `core/` to power the new `get_storage_capability` preflight endpoint. This matches the "environment access scattered outside config or bootstrap" red flag. The pattern is already widespread elsewhere in this component (not invented by this PR), but per this gate's rules a new call site in a previously-clean file still counts as extending the violation. Either thread `env` in as a parameter from the composition root, or record a written waiver for this component's established convention. See inline comment for detail.

Both are unwaived Invariant-classified violations (Rule 14), which is sufficient on its own for REQUEST_CHANGES regardless of the feature's size or urgency.

### Out-of-scope notes (not gating)
- The engine-bucket resolution reused via `resolve_baas_engine_bucket` (pre-existing registry in `core/bot_management/engines/registry.py`) correctly follows the contract+registry pattern — no new engine-string branching found in `core/common_config`.
- The `ArcaSandboxPlugin.create_sandbox` Plugin API extension (`volume_mounts`) is propagated to all four existing implementations (`_stub`, `aliyun_ack`, `local_docker`, `local_proc`) with explicit `NotImplementedError` for unsupported providers — correct plugin-lifecycle handling under Rule 11/25.
- No opinion offered on the rollout-whitelist semantics, retry/CAS correctness, or redaction logic in `bot_storage_policy.py` — those are business-logic concerns outside this gate's scope.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by totalfrank — APPROVED** [review]
## Architecture-Gate Review

**Verdict: APPROVE**

### Gate 0
- Target branch: `dev` (the integration branch) — REL-branch restrictions and the Isolation Audit do not apply. Gate 0 passes trivially.
- Classification: feature (`feat(backend): ...`), correctly targeted at `dev`.

### Review scope
Read `docs/arch/arch.rules.md` and `AGENTS.md`, then the full diff (47 files, both `src/backend` and `src/baas`). This PR introduces phase-1 UPFS storage-type support for BaaS-managed personal/service-Draft bot creation, gated by a config-driven rollout policy.

### Findings
None. Zero violations found in categories 1–11 of the violation catalog.

### Why this passes the engine-variability and plugin-boundary checks specifically

- **No engine-string branching added to core.** `upfs_rollout.should_rollout_upfs()` (`src/backend/.../core/common_config/upfs_rollout.py`) takes `engine` as an opaque string and matches it only against `engine_bucket` values sourced from external config (`bot_storage/upfs_rollout`) — no `if engine == "aicoding"`-style literals. The actual engine→bucket mapping is resolved once via the pre-existing `resolve_baas_engine_bucket` registry (`core/bot_management/engines/registry.py`, not touched by this PR) at the single call site in `baas_device_service.py`. Applying the "add engine `foo`" test: a new engine requires only its own strategy registration in the existing registry — zero edits to `common_config`/`upfs_rollout`/`bot_storage_policy`.
- **Contract + composition root pattern used correctly.** `BotStoragePolicyProtocol` / `BotCommonConfigServiceProtocol` are defined once (`core/common_config/bot_config_protocol.py`), implemented once (`bot_storage_policy.py`, `bot_config_service.py`), and wired only in composition roots (`di/modules/common_config_module.py`, `di/modules/service_bot_module.py`). `BaasService` depends only on the Protocol (`bot_storage_service: BotStoragePolicyProtocol | None = None`), with `None` as an explicit, documented default/opt-out — satisfying the "mandatory default implementation" requirement.
- **Plugin API extension done correctly.** `ArcaSandboxPlugin.create_sync_sandbox` (a true system-boundary Plugin API to Aliyun ACK/local sandboxes) gained an optional `volume_mounts` parameter, propagated to all four implementations (stub, local_docker, local_proc, aliyun_ack), each with explicit, documented behavior (aliyun_ack/local_docker/local_proc reject with `NotImplementedError`, matching the Protocol docstring's "unsupported implementations must reject them"). This is Rule 4-compliant explicit optional-behavior documentation, not silent capability drift.
- **No transport/env leakage introduced.** New core files have no web/RPC framework imports. `get_current_env()` calls added in `secbaas` core (`_device_template_service.py`) use the same `core.utils.env_utils` accessor this repo's own architecture test (`test_env_import_regression.py`) has already formally accepted as resolved Rule-14 debt (moved off `infra.utils` in 2026-06-26) — not a new boundary violation.
- **Config-driven, not hardcoded.** Rollout rules, UPFS subpath quota (`SystemConfigKey.UPFS_SUBPATH_SIZE_BYTES`), and per-template/env Volume IDs are all read from configuration/DB, not literals in core.
- **Migration is additive.** The new SQL (`2026_09_15_bot_common_config.sql`) only creates a new table; no existing schema is touched.
- **Propagation documented.** `common_config/README.md` and `docs/implementation/upfs-phase1.md` state the Service API change, consumers (BaasService, BaasDeviceService, publish/init task handlers), and explicit out-of-scope boundary (service Verify/Online, scaling, Caller, full migration — deferred). PR description mirrors this.

### Out-of-scope notes
- The production Aliyun ACK sandbox plugin currently rejects `volume_mounts` with `NotImplementedError`, so UPFS is not yet actually creatable against the real cloud provider even with rollout enabled — this is a stated phase-1 scope boundary (business completeness), not an architecture defect, and is consistent with the Plugin API's documented optional-capability contract.
- New Service/Plugin API protocols in this PR (`BotStoragePolicyProtocol`, `BotCommonConfigServiceProtocol`) were not added to `test_service_api_conformance.py`'s `_PAIRS` registry — this is consistent with existing precedent (the pre-existing sibling `CommonConfigServiceProtocol` in the same module is likewise absent), so not treated as a blocking gap here, but worth tightening in a follow-up.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by totalfrank — CHANGES_REQUESTED** [review]
## Architecture-Gate Review

**Verdict: REQUEST_CHANGES**

### Gate 0 — target branch

- Target: `dev` (integration branch), head `dev_upfs_phase1_20260915` @ `facddc3`.
- Classification: feature work (adds new Service API/protocol surface, DB table, HTTP endpoint, DTO fields).
- Result: **pass** — `dev` is the normal route for feature work; the Isolation Audit for release branches does not apply here.

This PR is otherwise a notably careful piece of engineering: engine variability is handled correctly (the new `BotStoragePolicyProtocol.initialize(engine=...)` treats `engine` as an opaque string matched against config-driven `engine_bucket` rules — no `if engine_type == "..."` branching was added anywhere in `core/`), the BaaS `ArcaSandboxPlugin` Protocol gets an additive `volume_mounts` parameter that every existing implementation must explicitly accept or reject (no silent ignore), storage-type selection flows through DTOs/enums rather than strings, DI wiring goes through composition roots (`di/modules/*`, `bootstrap/_core_services.py`), and new Service APIs ship with `README.md` Context Boundary declarations and a dedicated conformance-style test suite. I did not find any instance of catalog items 1–5, 7–11.

### Findings

**1. `src/baas/src/secbaas/community/core/service/template_manage/_device_template_service.py:170`** — new environment detection inside a core service method.

```python
def get_storage_capability(
    self, tenant: str, template_uuid: str, env: str | None = None,
) -> TemplateStorageCapability:
    current_env = get_current_env()
    result = TemplateStorageCapability(template_uuid=template_uuid, env=current_env, upfs_ready=False)
    if env is not None and env != current_env:
        return result.model_copy(update={"reason": "environment_mismatch"})
    ...
    if not capability_config.get_effective_upfs_volume_id(current_env):
```

`DefaultDeviceTemplateService` previously had no environment dependency at all; this PR is the one that introduces `from secbaas.community.core.utils.env_utils import get_current_env` into this file and uses it to decide business behavior (which environment's Volume ID field is authoritative, whether the caller's requested `env` is accepted). This is catalog item 6 / `arch.rules.md` **Rule 14 (Invariant)** and the constitution's explicit Structural Review Red Flag "environment access is scattered outside config or bootstrap": *"Configuration drives wiring (Rule 14); core receives values, it does not go looking for them."*

The PR's own backend (`agentclaw`) side gets this right for the equivalent problem — `di/modules/service_bot_module.py` resolves `storage_env=env_utils.get_current_env()` once, in the composition root, and passes it into `BaasService.__init__`. `DefaultDeviceTemplateService` (and its composition root in `bootstrap/_core_services.py` / `_container.py`) should do the same: inject the current environment (or an `EnvironmentProvider`-style Plugin/config value) through the constructor instead of calling the global getter from inside a core method.

**2. `src/baas/src/secbaas/community/core/service/paas/_facade.py:464`** — same violation, extending an existing one in the same method.

```python
expected_volume = (
    template_config.get_effective_upfs_volume_id(get_current_env())
    if template_config else None
)
```

`merge_with_template_and_detail` (or the enclosing method) already computes `env = get_current_env()` at line 402, earlier in the *same* method, for the pre-existing ARCA template-id resolution. The new UPFS block added by this PR doesn't even reuse that local `env` — it calls the global getter a second time. Per the review brief, "new code that extends an existing violation is itself a violation," and this is a clean example: the fix isn't just "reuse the local variable" (that only hides the smell), the method itself should receive environment as a passed-in value from its caller/composition root rather than resolving it internally. At minimum, thread the already-computed `env` into the UPFS branch rather than adding a third call site to a pattern the constitution flags by name.

Neither finding has a written waiver (rule/reason/risk/compensating-controls/owner/expiry) in the PR description, so per the Governance Addendum these unwaived Invariant violations block merge as-is.

### Out-of-scope notes

- Rollout targeting logic in `upfs_rollout.py` (percentage/whitelist by `engine_bucket`, tail-digit, user list) is business policy, not gated here.
- Redaction regex coverage in `_safe_creation_error` and the specific CAS retry counts (10 attempts) are implementation choices, not architecture.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by totalfrank — CHANGES_REQUESTED** [review]
## Architecture-Gate Review

**Verdict: REQUEST_CHANGES**

### Gate 0 — Target branch

- Target: `dev` (the integration branch). `dev` is not a `REL\d{8}` release branch, so the release-branch gate (existing-behavior-fix / isolated-feature exception / Isolation Audit) does not apply. **Gate 0: PASS** — no re-target issue.

### Full architecture review

I read `docs/arch/arch.rules.md` and `AGENTS.md`, then walked the diff (47 files, +3069/-30) across both touched components: `src/backend/src/agentclaw/community/...` (Bot common-config store, storage-policy service, rollout decision, DI wiring, deploy composer) and `src/baas/src/secbaas/community/...` (device service, ARCA sandbox plugin contract, template capability, HTTP adapter).

Most of this change is well-layered:
- `core/common_config/bot_storage_policy.py` / `upfs_rollout.py`: rollout decisions are pure, config-driven (`rule.get("engine_bucket") != engine` compares against externally supplied config, not a hardcoded engine set) — no engine-string branching.
- The engine→bucket lookup used for the rollout key (`resolve_baas_engine_bucket`) reuses the pre-existing `engines/registry.py` composition root unmodified — correct reuse of the existing contract pattern rather than reinventing engine branching.
- `BotCommonConfigRepositoryProtocol` / `BotCommonConfigRepository`, `BotStoragePolicyProtocol` / `BotStoragePolicyService`, `BaasServiceProtocol.begin_bot_storage_creation`/`finish_bot_storage_creation`: proper Protocol-first contracts with matching implementations, wired only in `di/modules/*` composition roots.
- `env_utils.get_current_env()` is called in the `ServiceBotModule` composition root and threaded down as `storage_env` — correct per Rule 14.
- The `ArcaSandboxPlugin.create_sandbox` Plugin API contract gained `volume_mounts: list[VolumeMountSpec] | None = None` — additive, defaulted, and every existing implementation (`aliyun_ack`, `local_docker`, `local_proc`, `_stub`) was updated in lockstep; unsupported providers correctly raise `NotImplementedError` as their capability declaration.
- New SQL migration is `CREATE TABLE IF NOT EXISTS` only (net-new table); `StorageType.UPFS` is an additive enum member; `BotDeployContext.storage_type` is a defaulted new field. No existing signature was broken.
- The new `/storage-capability` HTTP router handler is a thin one-line delegation to the core service — no domain policy in the adapter.

### Findings

1. **`src/baas/src/secbaas/community/core/service/device_manage/_device_service.py:823-829, 1456-1462, 1605-1611`** — hardcoded, copy-pasted capability check duplicating a capability contract this same PR introduces.

   ```python
   if (
       deploy_config
       and deploy_config.storage
       and deploy_config.storage.type == "upfs"
       and provider_type != "ARCA"
   ):
       raise ValueError("UPFS storage requires an ARCA template")
   ```

   This exact block is pasted three times (device creation in `start_device`, `restart_device`, `update_device`), gating UPFS eligibility on a literal provider-name string comparison inside the generic `DefaultDeviceService` dispatcher. The PR already introduces the correct mechanism for this exact question — `DeviceTemplateManageService.get_storage_capability()` / `TemplateStorageCapability.upfs_ready`, a declared Service API capability contract keyed off the template's own declared readiness, additionally enforced at the Plugin API boundary where non-ARCA `ArcaSandboxPlugin` implementations already reject `volume_mounts` via `NotImplementedError`. Instead of asking that capability (or centralizing the check in one place), this core dispatcher independently re-implements the same knowledge as an inline string literal three times. This is exactly the "repeated cross-cutting logic … added by copy-paste instead of through an extension mechanism" red flag, and the "capability must come from the component's own declaration, not a hardcoded set" concern (arch.rules.md Rule 10 / Rule 1 — a capability decision now has two independent sources of truth that can drift). If a second UPFS-capable provider is ever added, three core call sites need to be found and edited by hand, with no test enforcing they stay in sync.

   **Required fix:** collapse the three copies into one private helper (e.g. `self._require_upfs_capable(provider_type)`), and prefer deriving the check from the same capability declaration the PR already built (`get_storage_capability`/`TemplateStorageCapability`) rather than a second, independent `!= "ARCA"` literal.

2. **Missing conformance-test coverage for the changed Plugin API contract (Rule 25).** `ArcaSandboxPlugin.create_sandbox` (`src/baas/.../spi/sandbox/arca/_protocols.py`) is a declared Plugin API and this PR changes its signature (`volume_mounts` parameter) across all four implementations. `src/baas/tests/unit/api/arca_sandbox/` contains only an empty `__init__.py` — there is no shared/parametrized conformance suite that exercises all implementations (including the `_stub` default) against the current contract, so nothing in CI would catch a future implementation silently accepting `volume_mounts` without honoring it, or a signature drift between implementations. The PR's own `test_upfs_storage.py` and DI smoke-test updates cover call sites, but not contract conformance across implementations. Given this PR is the one extending the contract, it should add (or the PR description should explicitly waive, per the Governance Addendum) a guard/conformance test asserting every `ArcaSandboxPlugin` implementation handles/rejects `volume_mounts` consistently with the documented contract.

### Out-of-scope notes

- `_device_service.py` already branches on `provider_type` (`SIGMA`/`LOCAL`/`POOLAB`/`TECLAW`/`K8S`/`DOCKER`/`ARCA`) extensively pre-existing this PR; I did not re-litigate that established pattern beyond the three new copies called out in Finding 1.
- Business semantics of the rollout rules (tail-digit bucketing, `allow_all_users`, template capability precheck) are outside this review's scope and were not evaluated.
- `BotStoragePolicyProtocol` declares a leading-underscore method (`_resolve_storage_policy`) as part of a public Protocol — unusual naming, not gated on here since it's a style question, not a layering violation.

---
_Generated by [Claude Code](https://claude.ai/code)_
